src/typecheck/mod.rs

```diff
@@ -1,38 +1,46 @@
+//! Typechecking and type inference.
 //!
-//! # Mode
+//! Nickel uses a mix of a bidirectional typechecking algorithm, together with standard
+//! unification-based type inference. Nickel is gradually typed, and dynamic typing is the default.
+//! Static typechecking is triggered by a type annotation.
 //!
-//! Typechecking can be made in to different modes:
-//! - **Strict**: correspond to traditional typechecking in statically typed languages. This
-//!   happens inside a statically typed block. Statically typed blocks are introduced by the type
-//!   ascription operator `:`, as in `1 + 1 : Num` or `let f : Num -> Num = fun x => x + 1 in ..`. This is
-//!   implemented by [`type_check_`] and variants.
-//! - **Non strict**: do not enforce any typing but continue to traverse the AST looking for other
-//!   typed blocks to typecheck and store the annotations of let bindings in the environment. This is
-//!   implemented by the [`walk`] function.
+//! # Modes
 //!
-//! The algorithm starts in non strict mode. It is switched to strict mode when entering a
-//! statically typed block (an expression annotated with a type), and is switched back to
-//! non-strict mode when entering an expression annotated with a contract. Type and contract
-//! annotations thus serve both another purpose beside enforcing a type or a contract, which is to
-//! switch the typechecking mode.
+//! Corresponding to dynamic typing static typing, the typechecking algorithm runs in two separate
+//! modes:
+//!
+//! - **enforce** corresponds to traditional typechecking in a statically typed language. This
+//!   happens inside a statically typed block. Such blocks are introduced by the type ascription
+//!   operator `:`, as in `1 + 1 : Num` or `let f : Num -> Num = fun x => x + 1 in ..`. Enforce
+//!   mode is implemented by [`type_check_`] and variants.
+//! - **walk** doesn't enforce any typing but traverses the AST looking for typed blocks to
+//!   typecheck. Walk mode also stores the annotations of bound identifiers in the environment. This
+//!   is implemented by the [`walk`] function.
+//!
+//! The algorithm starts in walk mode. A typed block (an expression annotated with a type) switches
+//! to enforce mode, and is switched back to walk mode when entering an expression annotated with a
+//! contract. Type and contract annotations thus serve as a switch for the typechecking mode.
+//!
+//! Note that the static typing part is based on the bidirectional typing framework, which defines
+//! two different modes. Thus, the enforce mode is itself divided again into **checking** mode and
+//! **inference** mode.
 //!
 //! # Type inference
 //!
 //! Type inference is done via a form of bidirectional typechecking coupled with unification, in
-//! the same spirit as GHC (Haskell), albeit the type system of Nickel is much simpler. The type of
+//! the same spirit as GHC (Haskell), albeit the type system of Nickel is simpler. The type of
```

Comment:
```suggestion
//! the same spirit as GHC (Haskell), although the type system of Nickel is simpler. The type of
```

Reply:
Out of curiosity, what is the rational here? Does albeit require a be+ing form after? Is there a nuance in the meaning? Or does it just sound better to you?

Reply:
Initially it was just because it intuitively sounded better, but I just looked it up & apparently "albeit" should only be used if the clause it introduces is *not* an independent clause. (See [here](https://www.scribbr.com/definitions/albeit/)). In this case "the type system of Nickel is simpler" is an independent clause - i.e. it would be a valid sentence on its own - so "although" should be used instead of "albeit" here.

But I'm not gonna pretend I knew any of that when I wrote the comment. :joy: